CCEffects - Add color channel offset effect, fix effect clipping artifacts

thayerandrews opened the pull request:
- Add CCEffectColorChannelOffset which allows the red, green, and blue color channels to be offset from one another by independent 2D position vectors.
- Fix clipping of effects like blur that increase the size of the affected sprite. The user can now specify padding for their effects that increases the size of the submitted geometry without changing the visible size of the sprite image. Enlarging the geometry adds scratch space around the sprite image so blurs, blooms, and other effects do not get clipped by the geometry edges.
- Update the OSX project to include all the new effects.

Commit:
CCEffectRenderer - Fix build warning

GLKMatrix4Invert() takes bool not BOOL as its final argument.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
CCEffectColorChannelOffset - Initial commit

Commit:
CCEffectColorChannelOffset - First pass at the implementation

The basics are working though alpha isn't handled yet. This also adds a test to CCEffectsTest.

Commit:
CCEffectColorChannelOffset - Compute output alpha as the average of the sample alpha values

Commit:
CCEffectsTest - Remove misplaced ' ' character

Commit:
CCEffect - Add support for effect padding

Fix clipping of effects that expand the visual footprint of the node they're applied to (blur for example).
The user can now specify horizontal and vertical padding that increases the size of the rendered
geometry but adjusts the texture coordinates such that the sprite image stays the same size. By
adding this padding, effects like blur no longer get clipped off by the edge of the sprite geometry.

Unfortunately, for sprites from sprite sheets, individual effects must also be modified to avoid
fetching outside the sprite's footprint within the sheet. Because GLES does not support the
clamp-to-border wrap mode, this must be handled within the shaders.

Commit:
CCEffectBlur - Add padding support to CCEffectBlur and add a test for it

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' into colorChannelSeparation

Conflicts:
	cocos2d-ios.xcodeproj/project.pbxproj
	cocos2d-ui-tests/tests/CCEffectsTest.m

Commit:
CCEffects - Add CCEffectColorChannelOffset

This was removed while resolving a merge conflict.

Commit:
Merge branch 'develop' into effectPadding

Conflicts:
	cocos2d-ui-tests/tests/CCEffectsTest.m

Commit:
CCEffectsColorChannelOffset - Add support for effect padding

Commit:
CCEffects - Don't apply name mangling to the texture coordinate center and extent uniforms

They're CCEffect built-ins (not effect implementation specific) so they are shared between stacked
effects and should not be mangled.

Commit:
CCEffectStack - Update the padding of an effect stack from its contained effects

If the padding of any contained effects is greater than the padding set directly on the
stack, use the contained effects value.

Commit:
CCEffectRenderer - Fix texture coordinates for intermediate render passes from sprite sheets

We had been using the sprite's original texture coordinates for all render passes which will not
work with sprite sheets. When reading from an intermediate render target we need the texture
coordinates to be lower left (0,0) and upper right (1,1) and this is not going to be the case when
a sprite draws its image from a sprite sheet.

Fix this problem by overwriting the sprite texture coordinates with the required values when reading
from a sprite sheet.

Commit:
CCEffectsTest - Add labels to the sprites in the padding test

Commit:
CCEffects - Add color channel offset, drop shadow, and outer glow to the OSX Xcode project

slembcke commented:
Looks ok, but I can't run the tests. Did you forget to commit `Images/WhiteDiamond.png`?

Commit:
CCEffectsTest - Switch test image and effect for padding test

I don't want to introduce yet another test image.

thayerandrews commented:
Thanks for catching that. I don't actually want to add WhiteDiamong.png since I'm trying to avoid adding new test images if I don't need to. I'll update the test to use an existing file from the Images folder.

thayerandrews commented:
Okay. The tests should be fixed.

Commit:
CCEffectHue - Remove non-color matrix version of the hue transform

Commit:
CCEffects - Improve padding support

Effects can now inform the effect renderer how they are using the different texture coordinate sets
so the renderer can correctly decide how to transform the texture coordinates when padding is
requested.

Commit:
CCEffectBloom - Add support for padding to the bloom effect

Commit:
CCEffects - Adapt the default input snippets to effect padding

When effects that do not require padding are stacked with effects that do, their
input snippets need to handle the possibility that they will fetch out of bounds texels.

Commit:
CCEffectColorChannelOffset - Fix effect function name

Commit:
CCEffectBlur - Adapt blur to the recent effect renderer changes

Commit:
CCEffectsTest - Add more test cases to the padding test, add padding to the bloom test

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

thayerandrews commented:
**Update:**
- Fix padding support for effects like Bloom that use different texture coordinate sets to index the previous pass texture and the main texture.
- Fix padding support for effects that do not require padding themselves but have to account for padding when they are stacked with effects that do need it.
- Add more padding test cases.

slembcke commented on cocos2d/CCEffectBloom.m:
```diff
@@ -205,19 +217,27 @@ -(void)buildFragmentFunctions
             GLfloat optimizedOffset = (firstWeight * (currentOverlowTextureRead * 2 + 1) + secondWeight * (currentOverlowTextureRead * 2 + 2)) / optimizedWeight;
 
-            [shaderString appendFormat:@"srcPixel = texture2D(cc_PreviousPassTexture, v_blurCoordinates[0] + singleStepOffset * %f);\n", optimizedOffset];
+            [shaderString appendFormat:@"blurCoords = v_blurCoordinates[0] + singleStepOffset * %f);\n", optimizedOffset];
```
This line outputs shader code with an unmatched parenthesis:

```
blurCoords = v_blurCoordinates[0] + singleStepOffset * 7.384912);
```

slembcke commented on cocos2d/CCEffectBloom.m:
```diff
@@ -205,19 +217,27 @@ -(void)buildFragmentFunctions
             GLfloat optimizedWeight = firstWeight + secondWeight;
```
Does it make sense to have separate independent/dependent texture read paths now that they are always dependent?

thayerandrews replied:
This is something I've been wondering about myself. I think we should test
the performance of the two different paths and see if it's worth
maintaining different implementations.

On Wed, Sep 17, 2014 at 5:40 PM, Scott Lembcke notifications@github.com
wrote:

> In cocos2d/CCEffectBloom.m:
> 
> > @@ -205,19 +217,27 @@ -(void)buildFragmentFunctions
> >              GLfloat optimizedWeight = firstWeight + secondWeight;
> 
> Does it make sense to have separate independent/dependent texture read
> paths now that they are always dependent?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/cocos2d/cocos2d-swift/pull/963/files#r17696250.

slembcke commented:
Looks like there is a minor issue with the bloom effect left.

slembcke commented:
Eh, ignore me. Had I looked at it for 10 more seconds before posting I would have come to the conclusion that the extra parenthesis should be removed. Will close this shortly.
